xds/internal/balancer/cdsbalancer/cluster_handler_test.go

```diff
@@ -0,0 +1,390 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cdsbalancer
+
+import (
+	"context"
+	"testing"
+
+	xdsclient "google.golang.org/grpc/xds/internal/client"
+	"google.golang.org/grpc/xds/internal/testutils/fakeclient"
+)
+
+const (
+	edsService = "EDS Service"
+	logicalDNSService = "Logical DNS Service"
+	edsService2 = "EDS Service 2"
+	logicalDNSService2 = "Logical DNS Service 2"
+	aggregateClusterService = "Aggregate Cluster Service"
+)
+
+// setupTests creates a clusterHandler with a fake xds client for control over xds client.
+func setupTests(t *testing.T) (*clusterHandler, *fakeclient.Client) {
+	xdsC := fakeclient.NewClient()
+	ch := &clusterHandler{
+		xdsClient: xdsC,
+		// This is will be how the update channel is created in cds. It will be a separate channel to the buffer.Unbounded.
+		// This channel will also be read from to test any cluster updates.
+		updateChannel: make(chan clusterHandlerUpdate, 1),
+	}
+	return ch, xdsC
+}
+
+// Simplest case: the cluster handler receives a cluster name, handler starts a watch for that cluster, xds client returns
+// that it is an EDS Cluster, not a tree, so expectation that update is written to buffer which will be read by CDS LB.
+func (s) TestUpdateRootClusterEDSSuccess(t *testing.T) {
+	ch, fakeClient := setupTests(t)
+	// When you first update the root cluster, it should hit the code path which will start a cluster node for that root.
+	// Updating the root cluster logically represents a ping from a ClientConn.
+	ch.updateRootCluster(edsService)
+
+	// Starting a cluster node involves communicating with the xdsClient, telling it to watch a cluster.
+	ctx, ctxCancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer ctxCancel()
+	gotCluster, err := fakeClient.WaitForWatchCluster(ctx)
+	if err != nil {
+		t.Fatalf("xdsClient.WatchCDS failed with error: %v", err)
+	}
+	if gotCluster != edsService {
+		t.Fatalf("xdsClient.WatchCDS called for cluster: %v, want: %v", gotCluster, edsService)
+	}
+
+	// Invoke callback with xds client with a certain clusterUpdate. Due to this cluster update filling out the whole
+	// cluster tree, as the cluster is of type EDS and not an aggregate cluster, this should trigger the ClusterHandler
+	// to write to the update buffer to update the CDS policy.
+	fakeClient.InvokeWatchClusterCallback(xdsclient.ClusterUpdate{
+		ClusterType: xdsclient.ClusterTypeEDS,
+		ServiceName: edsService,
+	}, nil)
+
+	select {
+		case chu := <-ch.updateChannel:
+			if len(chu.chu) != 1 {
```

Comment:
Nit: you can compare the update slice with a wanted slice, and it will cover the length and each individual fields `cmp.Diff(chu.chu, want)`.

e.g. https://github.com/grpc/grpc-go/blob/master/xds/internal/balancer/clusterimpl/balancer_test.go#L168
(I believe you don't need the `cmpOpts` because all fields in `ClusterUpdate` are exported. Ping me if you get problems)

Reply:
Compared with cmp.Diff in every test :).